### PR TITLE
(DOCSP-27550): Kotlin Define an Object Model

### PR DESCRIPTION
**Pull request by cbullinger**
## Pull Request Info
Blocks https://github.com/mongodb/docs-realm/pull/2514 ([DOCS-15613](https://jira.mongodb.org/browse/DOCSP-27550))

### Jira

- https://jira.mongodb.org/browse/DOCSP-27550 

### Staged Changes

- (new page) [Define an Object Model - Kotlin SDK](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-27550-define-obj-model/sdk/kotlin/realm-database/define-realm-object-model/)
- (changes) [Open & Close a Realm](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-27550-define-obj-model/sdk/kotlin/realm-database/realm-files/open-and-close-a-realm)
- (small changes) [Index a Field](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-27550-define-obj-model/sdk/kotlin/realm-database/schemas/index/)
- (small updates) [Primary Keys](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-27550-define-obj-model/sdk/kotlin/realm-database/schemas/primary-key/)
- (small updates) [Ignore a Field](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/docsp-27550-define-obj-model/sdk/kotlin/realm-database/schemas/ignore/)

### Reminder Checklist

If your PR modifies the docs, you might need to also update some corresponding
pages. Check if completed or N/A.

- [x] Create Jira ticket for corresponding docs-app-services update(s), if any
- [x] Checked/updated Admin API
- [x] Checked/updated CLI reference

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/kotlin/realm-database/define-realm-object-model.txt [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+--------------------
+
+Use annotations to add functionality to properties in your Realm object models.  
+
+Required and Optional Properties
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Kotlin, value types are implicitly non-nullable. 
+You can make properties optional (nullable) by appending
+`?  <https://kotlinlang.org/docs/null-safety.html>`__.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/define-realm-object-model.txt:102]: I'm torn on whether we think there's really value in taking devs out of our docs to go check out the Kotlin docs on null safety. Do we think this is a concept that Kotlin SDK devs need to be linked to?

If we do want to keep this link here, I think it should be on more than just the `?`. That looks really odd to me, and also isn't a very large area for a link. If you want to keep the link, I might suggest moving it to the "(nullable)" or explicitly stating something like "For more information on null safety in Kotlin, refer to the Kotlin docs..."

File: source/sdk/kotlin/realm-database/schemas/ignore.txt [hunk 2]
```diff
@@ -10,11 +10,14 @@ Ignore a Field - Kotlin SDK
...
 `@Ignore <{+kotlin-local-prefix+}io.realm.kotlin.types.annotations/-ignore/index.html>`__
 annotation:
 
 .. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.ignore.kt
    :language: kotlin
-   :copyable: false
+
+Ignored properties behave exactly like normal properties, but they are 
+not stored to the database.  
+They can't be used in queries and won't trigger Realm notifications.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/schemas/ignore.txt:23]: Nit: The style guide suggests no future tense, but this is a teeny nit.

```suggestion
They can't be used in queries and don't trigger Realm notifications.
```

File: source/sdk/kotlin/realm-database/schemas/primary-key.txt [hunk 3]
```diff
@@ -10,28 +10,32 @@ Primary Keys - Kotlin SDK
...
-- Primary key **values** are immutable. To change the primary key value of
+- Primary key values are immutable. To change the primary key value of
   an object, you must delete the original object and insert a new object
   with a different primary key value.
-
-- You cannot change the primary key **field** for an object type after
-  adding any object of that type to a realm.
+- Primary keys are nullable. ``null`` can only be the primary key of one object
+  in a collection.
+  Realm automatically indexes primary keys, so you can efficiently read and modify objects based on their primary key.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/schemas/primary-key.txt:38]: I _think_ this is meant to be its own item? It doesn't seem to directly follow from the nullable bullet.
```suggestion
- Realm automatically indexes primary keys, so you can efficiently read and modify objects based on their primary key.
```

**Review by dacharyc — CHANGES_REQUESTED** [review]
Nice work! A handful of small tweaks and a few nits, but otherwise looking good.

File: source/sdk/kotlin/realm-database/define-realm-object-model.txt [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+Overview
+--------
+
+Realm applications model data as objects composed of field-value
+pairs that each contain one or more data types or other Realm objects.
+
+The Kotlin SDK memory maps Realm objects directly to native Kotlin 
+objects, so there's no need to use a special data access library (such as an 
+:wikipedia:`ORM <Object-relational_mapping>`).
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/define-realm-object-model.txt:21]: I don't feel strongly either way -- grabbed this from another sdk's version of this page and kept that ORM wiki link in. I'll remove the parens/link.

File: source/sdk/kotlin/realm-database/define-realm-object-model.txt [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+   
+  .. literalinclude:: /examples/generated/kotlin/SchemaTest.snippet.example-schema.kt
+    :language: kotlin
+
+.. _kotlin-realm-schema:
+
+Realm Schema
+~~~~~~~~~~~~
+
+A **realm schema** is a list of valid object schemas that a realm may contain.
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/define-realm-object-model.txt:54]: I know exactly what you mean. I waffled a lot on how to organize this topic and what to include in it (and I think I'll end up revisiting it in the near future once I get some of the other outstanding kotlin content done). I'll try it as an include.

File: source/sdk/kotlin/realm-database/define-realm-object-model.txt (hunk 1 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/define-realm-object-model.txt:102]: I copied this from the Flutter doc, but agree now that you've pointed it out. Updated to fix the link text.

File: source/sdk/kotlin/realm-database/schemas/primary-key.txt (hunk 3 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/schemas/primary-key.txt:38]: Yuuuuup. Thanks for catching

File: source/sdk/kotlin/realm-database/define-realm-object-model.txt [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+and that it meets all of the constraints specified in the schema.
+
+.. _kotlin-define-a-new-object-type:
+
+Define a New Object Type
+------------------------
+
+To define a Realm object type: 
+
+1. Create a uniquely named Kotlin class that implements the
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/define-realm-object-model.txt:69]: Technically you [don't hyphenate when the adverb ends in _-ly_](https://prowritingaid.com/grammar/1000138/How-to-use-hyphen-after-adverbs#:~:text=When%20the%20adverb%20ends%20in,%2C%20adjectives%2C%20and%20other%20adverbs.)

File: source/sdk/kotlin/realm-database/realm-files/open-and-close-a-realm.txt [hunk 7]
```diff
@@ -25,6 +25,18 @@ and pass the resulting :file:`RealmConfiguration` to
    :language: kotlin
    :copyable: false
 
+Realm Schema
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/realm-files/open-and-close-a-realm.txt:28]: Just a thought - I wonder if this makes more sense above the "Open" section here as a top-level section? It also applies to the in-memory realm stuff.

File: source/sdk/kotlin/realm-database/define-realm-object-model.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (dacharyc) [reply source/sdk/kotlin/realm-database/define-realm-object-model.txt:62]: Ugh, really? I either never learned or had forgotten that rule! Thanks 😀

**Review by dacharyc — APPROVED** [review]
Looks great - nice work!

File: source/sdk/kotlin/realm-database/realm-files/open-and-close-a-realm.txt (hunk 7 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/realm-files/open-and-close-a-realm.txt]: Agreed. I moved it above that section. And, honestly, there's quite a bit missing from that Open a Realm page that I've noted as part of the kotlin docs audit, so I'll be revisiting it soon.

**Pull request merged**